xcube/util/zarrinsert.py

```diff
@@ -0,0 +1,126 @@
+# of the Software, and to permit persons to whom the Software is furnished to do
+# so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import json
+import numpy as np
+import os
+import shutil
+import xarray as xr
+
+from xcube.util.timecoord import get_time_in_days_since_1970
+
+
+def check_append_or_insert(time_range, output_path):
+    t1, t2 = time_range
+    if t1 != t2:
+        t_center = (t1 + t2) / 2
+    else:
+        t_center = t1
+    ds = xr.open_zarr(output_path)
+    if np.greater(t_center, get_time_in_days_since_1970(ds.time.values[-1])):
+        # append modus
+        return True
+    else:
+        return False
+
+
+def merge_single_zarr_into_destination_zarr(src_path, dst_path):
+    """Merging the data for the new time stamp into the existing and remaining zarr directory."""
+    merge = _check_if_unique(src_path, dst_path)
+    if merge:
+        ds_single = xr.open_zarr(src_path)
+        ds = xr.open_zarr(dst_path)
+        src_idx = 0
+        new_idx = np.amin(np.where(ds.time[:] > (ds_single.time.values[src_idx])))
+        ds.close()
+        ds_single.close()
+        # Preparing the source directory with the single time stamp to be ready for merging
+        # --> files of variables, excluding "lat" and "lon" need to be renamed
+        _rename_file(src_path, src_idx, new_idx)
+        # Preparing the destination directory to be ready for single time stam to be merged
+        # --> files of variables, excluding "lat" and "lon" need to be renamed
+        # The renaming needs to happen in reversed order and starting at the index of nearest above value:
+        for i in reversed(range(new_idx, ds.time.shape[0])):
+            _rename_file(dst_path, i, (i + 1))
+        # Final step: copy the single time stamp files into the destination zarr and adjusting .zarray to the change.
+        _copy_into_target(src_path, dst_path, new_idx)
+        return True
+    else:
+        return False
+
+
+def _check_if_unique(src_path, dst_path):
+    """Check if to be added time stamp is unique """
+    ds = xr.open_zarr(dst_path)
+    ds_single = xr.open_zarr(src_path)
+    mask = np.equal(ds.time.values, ds_single.time.values)
+    ds.close()
+    ds_single.close()
+    if not mask.any():
+        return True
+    else:
+        return False
+
+
+def _rename_file(path_to_ds, old_index, new_time_i):
```

Comment:
This implementation does not recognize if nested zarr is used. See https://zarr.readthedocs.io/en/stable/api/storage.html#zarr.storage.NestedDirectoryStore